File: OJ/LeetCode/src/0071 - Simplify Path [Stack].cs
using System;
using System.Collections.Generic;

public class Solution
{
    public string SimplifyPath(string path)
    {
        string[] elements = path.Split('/', StringSplitOptions.RemoveEmptyEntries);
        Stack<string> list = new Stack<string>();
        foreach (var s in elements)
        {
            if (s == ".") continue;
            if (s == "..")
            {
                if (list.Count == 0) continue;
                list.Pop();
            }
            else
                list.Push(s);
        }

        string res = "";
        while (list.Count != 0)
            res = "/" + list.Pop() + res;

        return res == "" ? "/" : res;
    }
}

File: LeetCode/src/0071 - Simplify Path [Stack].cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

class Solution {
public:
    vector<string> split(string s, char c) {
        vector<string> ans;
        string cur;
        for (char ch : s) {
            if (ch == c) {
                if (cur != "") {
                    ans.emplace_back(cur);
                    cur.clear();
                }
            }
            else
                cur += ch;
        }
        if (cur != "")
            ans.emplace_back(cur);

        return ans;
    }
    string simplifyPath(string path) {
        vector<string> elements = split(path, '/');
        stack<string> list;
        for (auto s : elements) {
            if (s == ".") continue;
            if (s == "..") {
                if (list.empty()) continue;
                list.pop();
            }
            else
                list.push(s);
        }

        string res = "";
        while (!list.empty()) {
            res = "/" + list.top() + res;
            list.pop();
        }

        return res == "" ? "/" : res;
    }
};